servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CallableSingle.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright © 2019 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Objects;
+import java.util.concurrent.Callable;
+
+import static io.servicetalk.concurrent.Cancellable.IGNORE_CANCEL;
+import static io.servicetalk.concurrent.internal.SubscriberUtils.handleExceptionFromOnSubscribe;
+
+final class CallableSingle<T> extends AbstractSynchronousSingle<T> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CallableSingle.class);
+    private final Callable<T> callable;
+
+    CallableSingle(final Callable<T> callable) {
+        this.callable = Objects.requireNonNull(callable);
+    }
+
+    @Override
+    void doSubscribe(final Subscriber<? super T> subscriber) {
+        try {
+            subscriber.onSubscribe(IGNORE_CANCEL);
```

Comment:
Since the call to `Callable` may block, we should make sure that the calling thread is interrupted if the `Cancellable` is cancelled. We have a utility `ThreadInterruptingCancellable`[1] to handle such scenarios and you can refer to `BlockingUtils#blockingToSingle()`[2] on how to use it. Let me know if you need help on this.

Once you do this change, please also add a unit test to validate.

[1] https://github.com/apple/servicetalk/blob/master/servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ThreadInterruptingCancellable.java#L31
[2] https://github.com/apple/servicetalk/blob/master/servicetalk-http-api/src/main/java/io/servicetalk/http/api/BlockingUtils.java#L68

Reply:
The test is a bit specific to this case, so I wrote a sample for you:

```java
@Test
public void cancelInterrupts() throws Exception {
    final Single<Integer> source = Single.fromCallable(factory);
    CountDownLatch latch = new CountDownLatch(1);
    when(factory.call()).then(invocation -> {
        try {
            // await till interrupted.
            latch.await();
        } catch (InterruptedException e) {
            latch.countDown(); // signal that we got interrupted for the test to proceed
        }
        return 1;
    });
    toSource(source).subscribe(new SingleSource.Subscriber<Integer>() {
        @Override
        public void onSubscribe(final Cancellable cancellable) {
            cancellable.cancel();
        }

        @Override
        public void onSuccess(@Nullable final Integer result) {
            // noop
        }

        @Override
        public void onError(final Throwable t) {
            // noop
        }
    });
    latch.await();
}
```

The complexity here is that `subscribe()` will not return as `Callable#call()` blocks but we need access to the `Cancellable` to cancel. Let me know if you need clarification on the code.